README.md
# Version Logicielle pour toutes les versions du matériel

Cette version logicielle est compatible avec la version matérielle [1.2][1] ainsi que les anciennes versions des cartes electroniques. Elle est aussi compatible avec les cartes à base d'ESP8266 via un [adaptateur][5]

## Configuration

### Particle
-----------------------

- Télécharger l'environnement de développement [Particle-Dev][2] puis lancer l'IDE.
- Télécharger l'archive du repo logiciel [github][3] et le copier sur votre disque dur puis le décompresser
- Une fois l'IDE ouvert, menu File/Open et ouvrir le dossier programmateur-fil-pilote-wifi-master\Logiciel\remora
- Dans le menu Particle/Log in to Particle Cloud, entrez vos identifiants Particle pour vous connecter
- Dans le menu Particle/Select Device selectionnez votre Spark Core ou Particle Photon
- Ouvrir ensuite depuis l'IDE les fichers remora.ino et remora.h
- Selectionner la version de carte utilisé dans le fichier remora.h (les defines REMORA_BOARD_Vxx)
- Selectionner les modules utilisés dans le fichier remora.h (les defines MOD_xxxx)
- Selectionner l'onglet remora.ino
- Lancer la compilation+upload (icone éclair en haut à gauche)
- Si vous avez une erreur de type "App code was invalid" durant la compilation, supprimer le ficher README.md du dossier

### Arduino ESP8266 NodeMCU / Wemos d1 mini
--------------------------------------

- Suivre la procédure détaillée [ici](https://github.com/esp8266/Arduino), Attention ~~Arduino 1.6.5 pas 1.6.6~~ Maintenant c'est compatible Arduino 1.8.3 et je vous conseille d'utiliser la version git
- Installer le plugins SPIFFS pour l'IDE Arduino comme indiqué [ici](https://github.com/esp8266/arduino-esp8266fs-plugin)
- Lancer l'IDE Arduino
- Installer la librairie [NeoPixelBus](https://github.com/Makuna/NeoPixelBus) depuis l'IDE Menu Sketch / Include Library / Manage Library => NeoPixelBus by Makuna
- Installer la librairie [Blynk](https://github.com/blynkkk/blynk-library) depuis l'IDE Menu Sketch / Include Library / Manage Library / Blynk by Volodymyr Shymanskyy
- Installer la bibliothèque [ESPAsyncTCP](https://github.com/me-no-dev/ESPAsyncTCP) depuis l'IDE. Télécharger la dernière release au format *zip*, puis Menu **Croquis** / **Inclure une bibliothèque** / **Ajouter la bibliothèque .ZIP**
- Installer la bibliothèque [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer) depuis l'IDE. Télécharger la dernière release au format *zip*, puis Menu **Croquis** / **Inclure une bibliothèque** / **Ajouter la bibliothèque .ZIP**
- Installer la bibliothèque [esp8266-oled-ssd1306 v4.0.0](https://github.com/ThingPulse/esp8266-oled-ssd1306/releases/tag/4.0.0) depuis l'IDE Menu Sketch / Include Library / Manage Library => **ESP8266 Oled Driver for SSD1306 display**
- Lancer ensuite l'IDE Arduino et ouvrir depuis celui-ci le ficher `remora_soft.ino`
- Selectionner la version de carte utilisé dans le fichier [remora.h](https://github.com/hallard/remora_soft/blob/master/remora.h#L22-L26) (les defines REMORA_BOARD_Vxx)
- Selectionner les modules utilisés dans le fichier [remora.h](https://github.com/hallard/remora_soft/blob/master/remora.h#L28-L32) (les defines MOD_xxxx)
- choisir dans l'IDE Arduino Menu Tools / Board => **NodeMCU 1.0 (ESP12E-Module)** ainsi que le bon serial port
- choisir dans l'IDE Arduino Menu Tools / CPU Frequency => **160MHz**
- choisir dans l'IDE Arduino Menu Tools / Flash Size => **4M (1M SPIFFS)**
- mettre votre SSID et mot de passe Wifi dans le fichier remora.h.
`DEFAULT_WIFI_AP_PASS` est le mot de passe de connection Wifi au remora quand celui ci se positionne en point d'accès (lorsqu'il n'arrive pas à se connecter à vôtre Wifi par exemple) son IP est alors 192.168.4.1
````arduino
		// Définir ici les identifiants de
		// connexion à votre réseau Wifi
		// =====================================
		#define DEFAULT_WIFI_SSID "VotreSSID"
		#define DEFAULT_WIFI_PASS "VotreClé"
		#define DEFAULT_WIFI_AP_PASS "Remora_WiFi"
		// =====================================
````
- Lancer l'upload SPIFFS (les fichiers WEB) Menu Tools / ESP8266 Sketch Data Upload (c'est assez long)
- Lancer la compilation + upload (CTRL-U)

**Attention**, pour pouvoir uploader via le cable USB, la téléinfo ne doit pas être connectée (en OTA pas de soucis avec ça) car le port série est partagé. Donc soit il faut la débrancher soit il faut sortir le module NodeMCU du support.

Pour les mêmes raisons, et afin d'afficher les informations de debug dans une console, celle-ci doit être configurée à la même vitesse que la téléinfo (car c'est le même port série) donc 1200 bps, parité paire et 7 bits de data, or **cette configuration n'est pas possible dans le moniteur série de l'Arduino**. Il faut alors prendre un autre terminal comme putty, coolterm, ...

- La procédure OTA (télédéversement sans fil) est disponible [ici][8]


## Installation

Voici un [tutoriel d'installation de la Remora][9] qui vous permettra d'installer correctement votre **Remora** et de la configurer lors de votre premier accès.

## API Exposées

### Particle
------------

Toutes les API se font via des requêtes HTTP sur le cloud Particle.

A documenter!


### ESP8266 NodeMCU / Wemos d1 mini
-----------------------------------

Toutes les API se font via des requêtes HTTP sur le Remora. Il existe deux formats possibles si l'on veut récupérer des données ou exécuter des action avec le Remora. Chaque requête se verra retourner des données (ou un code de bonne éxécution) au format JSON.

Toute requête sera donc adressée sous la forme
`http://adresse_ip_du_remora/requete_plus_ou_moins_longue` dans les exemples ci dessous l'adresse IP de mon Remora est la 192.168.1.201, veillez à bien la changer pour mettre la vôtre. Les exemples ont été exécutés depuis la ligne de commande avec curl mais ils pourraient l'être depuis la barre d'addresse de votre navigateur.

#### Les Etats de fil pilote

Les différents états possibles de fil pilote dans l'API correspondent à la notation suivante, une lettre représente l'état lu ou le mode à positionner tel que :
```
C = Confort
A = Arrêt
E = Eco
H = Hors gel
1 = Eco-1 (non géré pour le moment)
2 = Eco-2 (non géré pour le moment)
```

#### Les Etats du mode de fonctionnement du relais

Voici les différents modes de fonctionnement du relais que vous pourrez trouver dans l'API:
```
0: arrêt
1: marche forcée
2: automatique
```

#### Les API d'intérrogation
----------------------------

Les API d'intérrogation se presentent sous la forme
`http://adresse_ip_du_remora/ma_donnee` et la/les donnée(s) sont retournées au format JSON (j'ai volontairement supprimé certains sauts de lignes de sortie pour une meilleure lecture)

- Durée de fonctionnement en secondes depuis le dernier reboot/reset/allumage `http://ip_du_remora/uptime`
````shell
    ~ # curl http://192.168.1.201/uptime
    { "uptime": 120 }
````
- Etat du relais et du mode de fonctionnement `http://ip_du_remora/relais`
````shell
    ~ # curl http://192.168.1.201/relais
    { "relais": 0, "fnct_relais": 2 }
````
- Etat du délestage `http://ip_du_remora/delestage`
````shell
		~ # curl http://192.168.1.201/delestage
		{ "niveau": 0, "zone": 1 }
````
Si le délestage est désactivé `http://ip_du_remora/delestage`
````shell
		~ # curl http://192.168.1.201/delestage
		{ "etat": "désactivé" }
````
- Etat d'un fil pilote `http://ip_du_remora/fpn` avec n = numéro du fil pilote (1 à 7)
````shell
		~ # curl http://192.168.1.201/fp3
		{ "fp3": "E" }
````
- Etat de tous les fils pilotes **http://ip_du_remora/fp**
````shell
		~ # curl http://192.168.1.201/fp
		{
		"fp1": "A",
		"fp2": "E",
		"fp3": "E",
		"fp4": "E",
		"fp5": "C",
		"fp6": "H",
		"fp7": "C"
		}
````
- Récupérer une étiquette Téléinfo par non nom `http://ip_du_remora/Nom_Etiquette`
````shell
		~ # curl http://192.168.1.201/PAPP
		{ "PAPP": 170 }
		~ # curl http://192.168.1.201/IINST
		{ "IINST": 1 }
		~ # curl http://192.168.1.201/PTEC
		{ "PTEC":"HC.." }
````
- Récupérer toutes les étiquettes Téléinfo en une fois `http://ip_du_remora/tinfo`
````shell
		~ # curl http://192.168.1.201/tinfo
		{ "_UPTIME":1614,
			"ADCO":31428067147,
			"OPTARIF":"HC..",
			"ISOUSC":15,
			"HCHC":410994,
			"HCHP":0,
			"PTEC":"HC..",
			"IINST":1,
			"IMAX":1,
			"PAPP":170,
			"HHPHC":3,
			"MOTDETAT":0
		}
````
A noter la présence de certaines étiquettes virtuelles commencant par un `_`


#### Les API d'action
---------------------

Les API d'action se presentent sous la forme
`http://adresse_ip_du_remora/?action=ma_donnee`, notez la différence avec les intérrogations, le `?`. Le résultat est retourné au format JSON avec un code réponse, il est :
- négatif en cas d'erreur
- à 0 si tout est OK
- positif pour indiquer un code retour OK différent si besoin.

Note, il est possible d'enchainer les actions en une requête mais un seul code d'erreur sera retourné pour l'ensemble, si une des commandes échoue, il faudra intérroger afin de savoir laquelle n'a pas fonctionnée.

##### Système
- Faire un reset (reboot) `http://ip_du_remora/reset`
````shell
		# curl http://192.168.1.201/reset
		OK, Redémarrage en cours
````

- Faire une remise à zero usine `http://ip_du_remora/factory_reset`, **attention tout va être vidé**, y compris le système de fichier SPIFFS, c'est comme si votre NodeMCU arrivait d'usine
````shell
		# curl http://192.168.1.201/factory_reset
		OK, Redémarrage en cours
````

##### Relais
- Activer le relais `http://ip_du_remora/?relais=1`
````shell
		# curl http://192.168.1.201/?relais=1
		{ "response": 0 }
````
- désactiver le relais `http://ip_du_remora/?relais=0`
````shell
		# curl http://192.168.1.201/?relais=0
		{ "response": 0 }
````
- Exemple d'erreur avec le relais `http://ip_du_remora/?relais=3`
````shell
		# curl http://192.168.1.201/?relais=3
		{ "response": -1 }
````
- Utiliser le mode automatique du relais basé sur les changements de périodes tarifaires
````shell
		# curl http://192.168.1.201/?frelais=2
		{ "response": 0 }
````
- Arrêter la gestion du relais et ouvrir le contact
````shell
		# curl http://192.168.1.201/?frelais=0
		{ "response": 0 }
````
- Mettre le relais en marche forcée
````shell
		# curl http://192.168.1.201/?frelais=1
		{ "response": 0 }
````
Il est aussi possible de forcer le relais jusqu'au prochain changement de période tarifaire, si le mode est en automatique. Pour cela, il vous suffit d'activer le relais directement.
````shell
		# curl http://192.168.1.201/?relais=1
		{ "response": 0 }
````

##### Fils pilotes
- selectionne le mode d'un des fils pilotes `http://ip_du_remora/?setfp=na` avec n=numéro du fil pilote et a=le mode à positionner (non sensible à la casse)
  Fil pilote 1 en arret
````shell
		# curl http://192.168.1.201?setfp=1a
		{ "response": 0 }
````
  Fil pilote 7 en Eco
````shell
		# curl http://192.168.1.201?setfp=7E
		{ "response": 0 }
````
  Mauvaise commande
````shell
		curl http://192.168.1.201?setfp=5X
		{ "response": -1 }
````
- Selectionne le mode d'un ou plusieurs les fils pilotes d'un coup `http://ip_du_remora/?fp=CMD` avec 7 commandes de fil pilote.
  Tous les fils pilote en confort
````shell
		curl http://192.168.1.201/?fp=CCCCCCC
		{ "response": 0 }
````
  Tous les fils pilote en arret
````shell
		curl http://192.168.1.201/?fp=AAAAAAA
		{ "response": 0 }
````
  Tous les fils pilote en eco
````shell
		curl http://192.168.1.201/?fp=EEEEEEE
		{ "response": 0 }
````
  Tous OFF sauf le fil pilote 1 en confort
````shell
		curl http://192.168.1.201/?fp=CAAAAAA
		{ "response": 0 }
````
  Tous OFF sauf le fil pilote 2 A-AAAAA
````shell
		curl http://192.168.1.201/?fp=CCCCCCC
		{ "response": 0 }
````
  FP1 Eco, FP2 inchangé, FP3 confort, FP4 hors gel, FP5 arrêt, FP6 Eco-1, FP7 Eco-2
````shell
		curl http://192.168.1.201/?fp=E-CHA12
		{ "response": -1 }
````
Erreur car les modes ECO-1 et ECO-2 ne sont pas gérés pour le moment.

## Afficheur OLED (Only Arduino ESP8266)
----------------------------------------

### Activation de l'afficheur

Pour activer l'afficheur sur votre **Remora**, il vous faut décommenter la variable **MOD_OLED** dans le fichier `remora.h`.

### Fonctionnement

L'afficheur affiche dorénavant plusieurs écrans en boucle:

 - Le logo de la Remora
 - La téléinfo
 - L'adresse IP WiFi
 - Les infos RF (Seulement si le **MOD_RF69** est activé)

Vous pouvez modifier la fréquence de changement d'affichage des frames avec la variable **DISPLAY_FPS** dans le fichier `display.h`.
Pour le moment, seuls les afficheurs OLED 128*64 sont gérés.

## A faire

- Mettre des icones plus sympas sur l'afficheur
- Gérer les retour de sondes RF, pour le moment seules les trames sont affichées sur la serial, rien n'est encore fait
- Gérer le RF des sondes Oregon Scientific
- pour le moment seul les [OLED][4] I2C sont gérés
- tout autre idée est bienvenue

## Historiques des Modifications

16/02/2015 : Ajout délestage cascadocyclique / Possibilité de ne récupérer l'état que d'un seul fil pilote

14/04/2015 : Ajout d'une variable spark pour la teleinfo. Passage en un seul appel pour script jeedom. Les variables d'origine restent utilisables.

30/09/2015 : voir le post [dédié][6].

02/12/2015 : Ajout [Version 1.3][7] Remora NodeMCU

04/01/2016 : Integration Interface WEB

01/10/2016 : Ajout de la (dés)activation du délestage. Ajout de la gestion automatisée du mode du relais.

07/01/2017 : Modification de la taille de la clé API pour Jeedom (32 => 48)

08/02/2017 : Ajout de la variable RELAIS_REVERSE à décommenter dans remora.h pour inverser l'état du relais en cas de problème

04/03/2017 : Ajout des connexions TCP Asynchrones

13/08/2017 : Affichage de plusieurs frames sur l'écran OLED

## Exemple

Pour les photos la téléinfo n'était pas branchée sur la carte, c'est pour celà que les compteurs et le bargraphe sont à 0

<img src="https://github.com/thibdct/programmateur-fil-pilote-wifi/blob/master/Logiciel/images/remora01.jpg" alt="Top"    width="45%" height="45%">&nbsp;
<img src="https://github.com/thibdct/programmateur-fil-pilote-wifi/blob/master/Logiciel/images/remora02.jpg" alt="Bottom" width="45%" height="45%">
<br>

Avec la téléinfo branchée sur la carte, le bargraph indique la puissance instantanée par rapport au contrat souscrit, ici soit 11% car on consomme 5A sur un contrat de 45A. Nous sommes en heures pleines (affichage sur fond blanc)
<img align="center" src="https://github.com/thibdct/programmateur-fil-pilote-wifi/blob/master/Logiciel/images/remora03.jpg" alt="Bottom" width="45%" height="45%">


[1]: https://github.com/thibdct/programmateur-fil-pilote-wifi/tree/master/Mat%C3%A9riel/1.2
[2]: https://www.particle.io/dev
[3]: https://github.com/hallard/remora_soft/archive/master.zip
[4]: http://www.ebay.com/itm/291216700457
[5]: https://github.com/hallard/Particle2NodeMCU/blob/master/README.md
[6]: https://community.hallard.me/topic/92/nouvelle-version-remora-compatible-esp8266
[7]: https://github.com/thibdct/programmateur-fil-pilote-wifi/tree/master/Mat%C3%A9riel/1.3
[8]: http://hallard.me/esp8266-ota/
[9]: https://www.aufilelec.fr/installation-remora/


CHANGELOG.md
# Change Log

## [v1.4.0](https://github.com/hallard/remora_soft/tree/v1.4.0) (2018-12-16)

[Full Changelog](https://github.com/hallard/remora_soft/compare/v1.3.5...v1.4.0)

**Implemented enhancements:**

- Ajout du display SSH1106 [\#38](https://github.com/hallard/remora_soft/pull/38) ([Azema](https://github.com/Azema))
- Add brightness LED RGB [\#35](https://github.com/hallard/remora_soft/pull/35) ([Azema](https://github.com/Azema))
- Display frames [\#32](https://github.com/hallard/remora_soft/pull/32) ([Azema](https://github.com/Azema))

**Fixed bugs:**

- Erreur de Push sur Jeedom avec une connexion SSL [\#33](https://github.com/hallard/remora_soft/issues/33)
- bouton RAZ usine sauvegarde la config [\#30](https://github.com/hallard/remora_soft/issues/30)
- Problème enregistrement configuration [\#26](https://github.com/hallard/remora_soft/issues/26)
- Correction compilation MOD\_OLED [\#40](https://github.com/hallard/remora_soft/pull/40) ([Azema](https://github.com/Azema))
- Update remora\_soft for OLEDisplay lib v4.0.0 [\#39](https://github.com/hallard/remora_soft/pull/39) ([Azema](https://github.com/Azema))
- Ajout init Emoncms [\#36](https://github.com/hallard/remora_soft/pull/36) ([Azema](https://github.com/Azema))
- Fix connexion SSL [\#34](https://github.com/hallard/remora_soft/pull/34) ([Azema](https://github.com/Azema))
- Correction d'erreur d'ID dans la page HTML [\#31](https://github.com/hallard/remora_soft/pull/31) ([Azema](https://github.com/Azema))

**Closed issues:**

- Problème de communication avec Jeedom   [\#29](https://github.com/hallard/remora_soft/issues/29)

## [v1.3.5](https://github.com/hallard/remora_soft/tree/v1.3.5) (2017-09-10)

[Full Changelog](https://github.com/hallard/remora_soft/compare/v1.3.4...v1.3.5)

**Fixed bugs:**

- Fixe problème d'enregistrement de config [\#27](https://github.com/hallard/remora_soft/pull/27) ([Azema](https://github.com/Azema))

**Closed issues:**

- Remontée info HC/HP via PTEC sous Jeedom [\#20](https://github.com/hallard/remora_soft/issues/20)
- Ordre des couleurs R et G pour les Led WS2812B [\#15](https://github.com/hallard/remora_soft/issues/15)

**Merged pull requests:**

- increment version to v1.3.5 [\#28](https://github.com/hallard/remora_soft/pull/28) ([Azema](https://github.com/Azema))
- FIX jeedom key size [\#25](https://github.com/hallard/remora_soft/pull/25) ([Azema](https://github.com/Azema))
- Tick Jeedom [\#24](https://github.com/hallard/remora_soft/pull/24) ([Azema](https://github.com/Azema))

## [v1.3.4](https://github.com/hallard/remora_soft/tree/v1.3.4) (2017-03-04)

[Full Changelog](https://github.com/hallard/remora_soft/compare/v1.3.3...v1.3.4)

**Merged pull requests:**

- Connexions TCP Async [\#22](https://github.com/hallard/remora_soft/pull/22) ([Azema](https://github.com/Azema))

## [v1.3.3](https://github.com/hallard/remora_soft/tree/v1.3.3) (2017-01-09)

[Full Changelog](https://github.com/hallard/remora_soft/compare/v1.3.2...v1.3.3)

**Merged pull requests:**

- Add jeedom custom port [\#19](https://github.com/hallard/remora_soft/pull/19) ([fmartinou](https://github.com/fmartinou))
- Jeedom key [\#16](https://github.com/hallard/remora_soft/pull/16) ([Azema](https://github.com/Azema))
- Correction d'erreurs du fonctionnement du relais [\#14](https://github.com/hallard/remora_soft/pull/14) ([Azema](https://github.com/Azema))

## [v1.3.2](https://github.com/hallard/remora_soft/tree/v1.3.2) (2016-10-05)

[Full Changelog](https://github.com/hallard/remora_soft/compare/0d5b77b2b515a7f054c0dd0d5055c2fa57b2b1e1...v1.3.2)

**Implemented enhancements:**

- Ajout de la gestion des zones dans webdev [\#3](https://github.com/hallard/remora_soft/pull/3) ([Azema](https://github.com/Azema))

**Closed issues:**

- Empty SPIFFS ? [\#1](https://github.com/hallard/remora_soft/issues/1)

**Merged pull requests:**

- Correction init debug serial [\#13](https://github.com/hallard/remora_soft/pull/13) ([Azema](https://github.com/Azema))
- Gestion automatisée du relais [\#12](https://github.com/hallard/remora_soft/pull/12) ([Azema](https://github.com/Azema))
- Ajout de la fonction d'activation du délestage [\#10](https://github.com/hallard/remora_soft/pull/10) ([Azema](https://github.com/Azema))
- Debug [\#9](https://github.com/hallard/remora_soft/pull/9) ([Azema](https://github.com/Azema))
- Fix requêtes interrogations [\#7](https://github.com/hallard/remora_soft/pull/7) ([Azema](https://github.com/Azema))
- Ajout de la copie des fonts dans le script de creation des spiffs [\#5](https://github.com/hallard/remora_soft/pull/5) ([Azema](https://github.com/Azema))
- Ajout de la classe loaded à body [\#4](https://github.com/hallard/remora_soft/pull/4) ([Azema](https://github.com/Azema))



* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tinfo.cpp
// **********************************************************************************
// Teleinfo management file for remora project
// **********************************************************************************
// Copyright (C) 2014 Thibault Ducret
// Licence MIT
//
// History : 15/01/2015 Charles-Henri Hallard (http://hallard.me)
//                      Intégration de version 1.2 de la carte electronique
//           13/04/2015 Theju
//                      Modification des variables cloud teleinfo
//                      (passage en 1 seul appel) et liberation de variables
//           15/09/2015 Charles-Henri Hallard Utilisation Librairie Teleinfo Universelle
// **********************************************************************************

#include "tinfo.h"

#ifdef MOD_TELEINFO
// Instanciation de l'objet Téléinfo
TInfo tinfo;
#endif

uint mypApp           = 0;
uint myiInst          = 0;
uint myindexHC        = 0;
uint myindexHP        = 0;
uint myimax           = 0;
uint myisousc         = ISOUSCRITE; // pour calculer la limite de délestage
char myPeriode[8]     = "";
char mytinfo[250]     = "";
char mycompteur[64]   = "";
float ratio_delestage = DELESTAGE_RATIO;
float ratio_relestage = RELESTAGE_RATIO;
float myDelestLimit   = 0.0;
float myRelestLimit   = 0.0;
int etatrelais        = 0; // Etat du relais
int fnctRelais        = 2; // Mode de fonctionnement du relais
int lastPtec          = PTEC_HP;

unsigned long tinfo_led_timer = 0; // Led blink timer
unsigned long tinfo_last_frame = 0; // dernière fois qu'on a recu une trame valide

ptec_e ptec; // Puissance tarifaire en cours

/* ======================================================================
Function: ADPSCallback
Purpose : called by library when we detected a ADPS on any phase
Input   : phase number
            0 for ADPS (monophase)
            1 for ADIR1 triphase
            2 for ADIR2 triphase
            3 for ADIR3 triphase
Output  : -
Comments: should have been initialised in the main sketch with a
          tinfo.attachADPSCallback(ADPSCallback())
====================================================================== */
void ADPSCallback(uint8_t phase)
{
  // Led Rouge
  LedRGBON(COLOR_RED);
  tinfo_led_timer = millis();

  // Monophasé
  if (phase == 0 ) {
    DebuglnF("ADPS");
  } else {
    DebugF("ADPS Phase ");
    Debugln('0' + phase);
  }

  // nous avons une téléinfo fonctionelle
  status |= STATUS_TINFO;
  tinfo_last_frame = millis();
}

/* ======================================================================
Function: DataCallback
Purpose : callback when we detected new or modified data received
Input   : linked list pointer on the concerned data
          value current state being TINFO_VALUE_ADDED/TINFO_VALUE_UPDATED
Output  : -
Comments: -
====================================================================== */
void DataCallback(ValueList * me, uint8_t flags)
{
  // Do whatever you want there
  Debug(me->name);
  Debug('=');
  Debug(me->value);

  //Debug(" Flags=0x");
  //DEBUG_SERIAL.print(flags, HEX);

  if ( flags & TINFO_FLAGS_NOTHING ) DebugF(" Nothing");
  if ( flags & TINFO_FLAGS_ADDED )   DebugF(" Added");
  if ( flags & TINFO_FLAGS_UPDATED ) DebugF(" Updated");
  if ( flags & TINFO_FLAGS_EXIST )   DebugF(" Exist");
  if ( flags & TINFO_FLAGS_ALERT )   DebugF(" Alert");
  Debugln();
  Debugflush();

  // Nous venons de recevoir la puissance tarifaire en cours
  // To DO : gérer les autres types de contrat
  if (!strcmp(me->name, "PTEC")) {
    // Récupération de la période tarifaire en cours
    strncpy(myPeriode, me->value, strlen(me->value));

    // Determination de la puissance tarifaire en cours
    // To DO : gérer les autres types de contrat
    if (!strcmp(me->value, "HP..")) ptec= PTEC_HP;
    if (!strcmp(me->value, "HC..")) ptec= PTEC_HC;
    if (!strcmp(me->value, "TH..")) ptec= PTEC_HP;

    //=============================================================
    //    Ajout de la gestion du relais aux heures creuses
    //=============================================================
    if (fnctRelais == FNCT_RELAIS_AUTO && lastPtec != ptec) {
      //Debug("PTEC: ");
      if (ptec == PTEC_HC) {
        //Debugln(" HC");
        relais("1");
      } else {
        //Debugln(" HP");
        relais("0");
      }
      lastPtec = (int)ptec;
    }
  }

  // Mise à jour des variables "cloud"
  if (!strcmp(me->name, "PAPP"))   mypApp    = atoi(me->value);
  if (!strcmp(me->name, "IINST"))  myiInst   = atoi(me->value);
  if (!strcmp(me->name, "HCHC"))   myindexHC = atol(me->value);
  if (!strcmp(me->name, "HCHP"))   myindexHP = atol(me->value);
  if (!strcmp(me->name, "IMAX"))   myimax    = atoi(me->value);
  if (!strcmp(me->name, "BASE"))   { myindexHP = atol(me->value); myindexHC = 0; }

  // Isousc permet de connaitre l'intensité max pour le delestage
  if (!strcmp(me->name, "ISOUSC")) {
    myisousc = atoi(me->value);
    myDelestLimit = ratio_delestage * myisousc;
    // Calcul de quand on déclenchera le relestage
    myRelestLimit = ratio_relestage * myisousc;

    // Maintenant on connait notre contrat, on peut commencer 
    // A traiter le delestage eventuel et si celui-ci 
    // n'a jamais été initialisé on le fait maintenant
    if ( timerDelestRelest == 0 )
      timerDelestRelest = millis();
  }

  Debugln();

  // nous avons une téléinfo fonctionelle
  status |= STATUS_TINFO;
}

/* ======================================================================
Function: NewFrame
Purpose : callback when we received a complete teleinfo frame
Input   : linked list pointer on the concerned data
Output  : -
Comments: -
====================================================================== */
void NewFrame(ValueList * me)
{
  char buff[32];

  // Light the RGB LED
  LedRGBON(COLOR_GREEN);
  tinfo_led_timer = millis();

  #if defined (ESP8266)
    //sprintf( buff, "New Frame (%ld Bytes free)", ESP.getFreeHeap() );
  #elif defined (SPARK)
    //sprintf( buff, "New Frame (%ld Bytes free)", System.freeMemory());
  #else
    //sprintf( buff, "New Frame");
  #endif
  //Debugln(buff);

  // Ok nous avons une téléinfo fonctionelle
  status |= STATUS_TINFO;
  tinfo_last_frame = millis();
}

/* ======================================================================
Function: NewFrame
Purpose : callback when we received a complete teleinfo frame
Input   : linked list pointer on the concerned data
Output  : -
Comments: it's called only if one data in the frame is different than
          the previous frame
====================================================================== */
void UpdatedFrame(ValueList * me)
{
  char buff[32];

  // Light the RGB LED (orange) and set timer
  LedRGBON(COLOR_ORANGE);
  tinfo_led_timer = millis();

  #if defined (ESP8266)
    //sprintf( buff, "Updated Frame (%ld Bytes free)", ESP.getFreeHeap() );
  #elif defined (SPARK)
    //sprintf( buff, "Updated Frame (%ld Bytes free)", System.freeMemory());
  #else
    //sprintf( buff, "Updated Frame");
  #endif
  //Debugln(buff);

  //On publie toutes les infos teleinfos dans un seul appel :
  sprintf(mytinfo,"{\"papp\":%u,\"iinst\":%u,\"isousc\":%u,\"ptec\":%u,\"indexHP\":%u,\"indexHC\":%u,\"imax\":%u,\"ADCO\":%u}",
                    mypApp,myiInst,myisousc,ptec,myindexHP,myindexHC,myimax,mycompteur);
  // Posibilité de faire une pseudo serial avec la fonction suivante :
  //Spark.publish("Teleinfo",mytinfo);

  // nous avons une téléinfo fonctionelle
  status |= STATUS_TINFO;
  tinfo_last_frame = millis();
}

/* ======================================================================
Function: tinfo_setup
Purpose : prepare and init stuff, configuration, ..
Input   : indique si on doit bloquer jusqu'à reception ligne téléinfo
Output  : false si on devait attendre et time out expiré, true sinon
Comments: -
====================================================================== */
bool tinfo_setup(bool wait_data)
{
  bool ret = false;

  Debug("Initializing Teleinfo...");
  Debugflush();

  #ifdef SPARK
    Serial1.begin(1200);  // Port série RX/TX on serial1 for Spark
  #endif

  // reset du timeout de detection de la teleinfo
  tinfo_last_frame = millis();

  // Init teleinfo
  tinfo.init();

  // Attacher les callback donc nous avons besoin
  #ifdef MOD_ADPS
    tinfo.attachADPS(ADPSCallback);
  #endif
  tinfo.attachData(DataCallback);
  tinfo.attachNewFrame(NewFrame);
  tinfo.attachUpdatedFrame(UpdatedFrame);

  // Doit-on attendre une ligne valide
  if (wait_data) {
    // ici on attend une trame complete ou le time out
    while ( !(status & STATUS_TINFO) && (millis()-tinfo_last_frame<TINFO_FRAME_TIMEOUT*1000)) {
      char c;
      // Envoyer le contenu de la serial au process teleinfo
      // les callback mettront le status à jour
      #ifdef SPARK
        if ( Serial1.available()) {
          c = Serial1.read();
          //Debug(c);
          //Debugflush();
          tinfo.process(c);
        }
      #else
        if (Serial.available()) {
          c = Serial.read();
          //Debug(c);
          //Debugflush();
          tinfo.process(c);
        }
      #endif
      
      _yield();
    }
  }

  ret = (status & STATUS_TINFO)?true:false;
  Debug("Init Teleinfo ");
  Debugln(ret?"OK!":"Erreur!");

  return ret;
}

/* ======================================================================
Function: tinfo_loop
Purpose : gestion des trames reçues par la librairie teleinfo
Input   : -
Output  : -
Comments: -
====================================================================== */
void tinfo_loop(void)
{
#ifdef MOD_TELEINFO
  char c;
  uint8_t nb_char=0;
  // Evitons les conversions hasardeuses, parlons float
  float fiInst = myiInst; 

  // on a la téléinfo présente ?
  if ( status & STATUS_TINFO) {
    // est ce que cela fait un moment qu'on a pas recu de trame
    if ( millis()-tinfo_last_frame>TINFO_FRAME_TIMEOUT*1000) {
      // Indiquer qu'elle n'est pas présente
      status &= ~STATUS_TINFO;
      Debugln("Teleinfo absente/perdue!");
    }

  // Nous n'avions plus de téléinfo
  } else  {
    // est ce que cela fait un moment qu'on a pas recu de data
    if ( millis()-tinfo_last_frame>TINFO_DATA_TIMEOUT*1000) {
      // Light the RGB LED (RED) and set timer with
      // 500ms more than classic blink
      LedRGBON(COLOR_RED);
      tinfo_last_frame = millis();
      tinfo_led_timer = millis();
      Debugln("Teleinfo toujours absente!");
    }
  }

  // Caractère présent sur la sérial téléinfo ?
  // On prendra maximum 8 caractères par passage
  // les autres au prochain tour, çà evite les
  // long while bloquant pour les autres traitements
  #ifdef SPARK
    while (Serial1.available() && nb_char<8) {
      c = (Serial1.read());
      tinfo.process(c);
      nb_char++;
    }
  #else
    while (Serial.available() && nb_char<8) {
      c = (Serial.read());
      tinfo.process(c);
      nb_char++;
    }
  #endif

  // Faut-il enclencher le delestage ?
  #ifdef MOD_ADPS
  //On dépasse le courant max?
  if (fiInst > myDelestLimit) {
    if ((millis() - timerDelestRelest) > 5000L)  {
      //On ne passe pas dans la boucle si l'on a délesté ou relesté une zone il y a moins de 5s
      //On évite ainsi de délester d'autres zones avant que le délestage précédent ne fasse effet
      delester1zone();
      timerDelestRelest = millis();
    }
  } else {
    // Un délestage est en cours (nivDelest > 0)
    // Le délestage/relestage de la dernière zone date de plus de 3 minutes
    // On attend au moins ce délai pour relester ou décaler
    // pour éviter les délestage/relestage trop rapprochés
    if (nivDelest > 0 && (millis() - timerDelestRelest) > 180000L) {
      //Le courant est suffisamment bas pour relester
      if (fiInst < myRelestLimit) {
        relester1zone();
        timerDelestRelest = millis();
      } else {
        // On fait tourner le délestage
        // ex : AVANT = "DDCEEEE" => APRES = "CDDEEEE"
        decalerDelestage();
        timerDelestRelest = millis();
      }
    }
  }
  #endif //ADPS active

  // Do we have RGB led timer expiration ?
  if (tinfo_led_timer && (millis()-tinfo_led_timer >= TINFO_LED_BLINK_MS)) {
      LedRGBOFF(); // Light Off the LED
      tinfo_led_timer=0; // Stop virtual timer
  }

#endif
}
